lua/kickstart/plugins/debug.lua

```diff
@@ -0,0 +1,69 @@
+return {
+  {
+    enabled = true,
+    config = function()
+      -- Optional debug adapter setup
+      --
+      -- To enable setup, change `disable = true` in the packer section
+      -- of the init.lua. You'll also want to copy this file into your
+      -- config at ~/.config/nvim/after/plugin/dap.lua
+
+      local dapui = require 'dapui'
+
+      require('mason-nvim-dap').setup {
+        -- Makes a best effort to setup the various debuggers with
+        -- reasonable debug configurations
+        automatic_setup = true,
+
+        -- You'll need to check that you have the required things installed
+        -- online, please don't ask me how to install them :)
+        ensure_installed = {
+          -- Update this to ensure that you have the debuggers for the langs you want
+          'delve',
+        },
+      }
+
+      -- You can provide additional configuration to the handlers,
+      -- see mason-nvim-dap README for more information
+      require('mason-nvim-dap').setup_handlers()
+
+      -- Basic debugging keymaps, feel free to change to your liking!
+      vim.keymap.set('n', '<F5>', require('dap').continue)
+      vim.keymap.set('n', '<F1>', require('dap').step_into)
+      vim.keymap.set('n', '<F2>', require('dap').step_over)
+      vim.keymap.set('n', '<F3>', require('dap').step_out)
+      vim.keymap.set('n', '<leader>b', require('dap').toggle_breakpoint)
+      vim.keymap.set('n', '<leader>B', function()
+        require('dap').set_breakpoint(vim.fn.input 'Breakpoint condition: ')
+      end)
+
+      -- Dap UI setup
+      -- For more information, see |:help nvim-dap-ui|
+      dapui.setup {
+        -- Set icons to characters that are more likely to work in every terminal.
+        --    Feel free to remove or use ones that you like more! :)
+        --    Don't feel like these are good choices.
+        icons = { expanded = '▾', collapsed = '▸', current_frame = '*' },
+        controls = {
+          icons = {
+            pause = '⏸',
+            play = '▶',
+            step_into = '⏎',
+            step_over = '⏭',
+            step_out = '⏮',
+            step_back = 'b',
+            run_last = '▶▶',
+            terminate = '⏹',
+          },
+        },
+      }
+
+      dap.listeners.after.event_initialized['dapui_config'] = dapui.open
+      dap.listeners.before.event_terminated['dapui_config'] = dapui.close
+      dap.listeners.before.event_exited['dapui_config'] = dapui.close
+
+      -- Install golang specific config
+      require('dap-go').setup()
```

Comment:
Pardon my ignorance, but will it be clear to folks where they should add the require().setup() for their own favorite languages?

Reply:
well you'll have to emulate what I've done here for golang -- i can list a few examples and the wiki entries to explore. I haven't done a bunch of DAP stuff for other languages

Reply:
if you add:

Reply:
here is my full call to nvim-dap with packer, for clarification:

```
	use { 'mfussenegger/nvim-dap',
		requires = {
			'theHamsta/nvim-dap-virtual-text', -- Show variable values in virtual text
			'jay-babu/mason-nvim-dap.nvim', -- Automatic DAP configuration
			'williamboman/mason.nvim',
		},
		config = function()
			vim.fn.sign_define('DapBreakpoint', { text = '', texthl = 'Error' })
			vim.fn.sign_define('DapBreakpointCondition', { text = 'לּ', texthl = 'Error' })
			vim.fn.sign_define('DapLogPoint', { text = '', texthl = 'Directory' })
			vim.fn.sign_define('DapStopped', { text = 'ﰲ', texthl = 'TSConstant' })
			vim.fn.sign_define('DapBreakpointRejected', { text = '', texthl = 'Error' })

			-- Automatically set up installed DAP adapters
			require('mason-nvim-dap').setup({ automatic_setup = true })
			require('mason-nvim-dap').setup_handlers()
			-- DAP virtual text --
			require('nvim-dap-virtual-text').setup()
		end
	}
```

Reply:
Right, but some of the plugins give you additional configuration that allow you to do different types of debugging (like attaching to running process, run individual unit tests, etc)

Reply:
This is something I've struggled with and hope to contribute back to help other newbies :)

Specifically, I find it tremendously confusing to understand how to express the configuration for any given plugin, and I think the root of the confusion is understaning what lua lexical/function closure/scope a given plugin wants for its configuration.

Take [this](https://github.com/feoh/git_dotfiles/blob/master/.config/nvim/lua/custom/plugins.lua#L1) for example - it took me forever to understand that wrapping all my extra 'use' directives in a `return function(use)` would allow Packer to ingest them despite their being in another file.

Is there a way to signpost this for new users? Will Lazy make this easier?

Thanks so much for all your hard work. kickstart.nvim & neovim have made me a more productive developer == PRICELESS :)

Reply:
yes, the new method is much simpler and less error prone. I will add docs as well before I merge